core/primitives/src/runtime/parameter_table.rs

```diff
@@ -16,36 +16,90 @@ pub(crate) enum ParameterValue {
     String(String),
 }
 
-impl ParameterValue {
-    fn as_u64(&self) -> Option<u64> {
-        match self {
-            ParameterValue::U64(v) => Some(*v),
-            _ => None,
+#[derive(thiserror::Error, Debug)]
+pub(crate) enum ValueConversionError {
+    #[error("expected a value of type `{0}`, but could not parse it from `{1:?}`")]
+    WrongType(&'static str, ParameterValue),
+
+    #[error("expected an integer of type `{1}` but could not parse it from `{2:?}`")]
+    ParseInt(#[source] std::num::ParseIntError, &'static str, ParameterValue),
+    // TODO(akashin): Enable this.
+    // #[error("expected an integer of type `{1}` but could not parse it from `{2:?}`")]
+    // TryFromInt(#[source] std::num::TryFromIntError, &'static str, ParameterValue),
+}
+
+impl TryFrom<&ParameterValue> for u64 {
+    type Error = ValueConversionError;
+
+    fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+        match value {
+            ParameterValue::U64(v) => Ok(*v),
+            _ => Err(ValueConversionError::WrongType(std::any::type_name::<u64>(), value.clone())),
         }
     }
+}
 
-    fn as_rational(&self) -> Option<Rational32> {
-        match self {
-            ParameterValue::Rational { numerator, denominator } => {
-                Some(Rational32::new(*numerator, *denominator))
-            }
-            _ => None,
+impl TryFrom<&ParameterValue> for u128 {
+    type Error = ValueConversionError;
+
+    fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
+        match value {
+            ParameterValue::U64(v) => Ok(u128::from(*v)),
+            ParameterValue::String(s) => s.parse().map_err(|err| {
+                ValueConversionError::ParseInt(err, std::any::type_name::<u128>(), value.clone())
+            }),
+            _ => Err(ValueConversionError::WrongType(std::any::type_name::<u128>(), value.clone())),
         }
     }
+}
 
-    fn as_u128(&self) -> Option<u128> {
-        match self {
-            ParameterValue::U64(v) => Some(u128::from(*v)),
-            // TODO(akashin): Refactor this to use `TryFrom` and properly propagate an error.
-            ParameterValue::String(s) => s.parse().ok(),
-            _ => None,
+// TODO(akashin): Enable this.
+// impl<T> TryFrom<&ParameterValue> for T
```

Comment:
This is one of those cases where implementing an `*Into` trait would make sense.

Reply:
I've tried
```rust
impl<T> TryInto<T> for &ParameterValue
where
    T: TryFrom<u64>,
    T::Error: Into<std::num::TryFromIntError>,
{
...
}
```

and this now fails with

```
error[E0119]: conflicting implementations of trait `std::convert::TryInto<_>` for type `&runtime::parameter_table::ParameterValue`
  --> core/primitives/src/runtime/parameter_table.rs:43:1
   |
43 | impl<T> TryInto<T> for &ParameterValue
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: conflicting implementation in crate `core`:
           - impl<T, U> TryInto<U> for T
             where U: TryFrom<T>;
```

This seems to be a known problem https://github.com/rust-lang/rust/issues/50133, but I haven't yet figured out how to use a workaround mentioned there https://github.com/rust-lang/rust/issues/50133#issuecomment-646908391.

Reply:
Ah, I confused myself here thinking the conversion was in the other direction than I was thinking of. What you could do is introduce a trait for types that can be extracted, and keep just this generic implementation, kinda like this:

```
trait ParameterValueContents { ... }
impl ParameterValueContents for u64 { ... }
...

impl<T: ParameterValueContents> TryFrom<...> for T { ... }
```

But that’s just yuck. I’d say just make a macro and implement the trait for all the interesting conversions directly.

```rust
macro_rules! implement_conversion_to {
    ($($ty: ty),*) => {
        $(impl TryFrom<&ParameterValue> for $ty {
            type Error = ValueConversionError;
            fn try_from(value: &ParameterValue) -> Result<Self, Self::Error> {
                match value {
                    ParameterValue::U64(v) => <$ty>::try_from(*v).map_err(|err| todo!("didn't fit")),
                    _ => todo!("..."),
                }
            }
        })*
    }
}

implement_conversion_to!(u128, u64, u32, u16, u8, i128, i64, i32, i16, i8, usize, isize);
```

Reply:
Wow, this worked! Thanks for helping to write my first macro in NEAR codebase :)